core/allocators.h

```diff
@@ -125,42 +117,41 @@ class BalloonAllocator {
 
 		Hand &h = hands[hand];
 
-		if (b == h.first)
-			h.first = b->next;
-
+		// Remove balloon
 		if (b->prev)
 			b->prev->next = b->next;
+		else
+			h.first = b->next;
+
 		if (b->next)
 			b->next->prev = b->prev;
 
-		if (h.last != b) {
-			h.last->next = b;
+		if (h.last != b) { // Place balloon last, so if can be reused
 			b->prev = h.last;
 			b->next = NULL;
+			h.last->next = b;
 			h.last = b;
 		}
 
 		h.used--;
-
-		if (h.used <= (h.allocated - (PREALLOC_COUNT * 2))) { // this is done to ensure no alloc/free is done constantly
+		if (h.used <= (h.allocated - (PREALLOC_COUNT * 2))) { // Used to free up memory
 
 			for (int i = 0; i < PREALLOC_COUNT; i++) {
 				ERR_CONTINUE(h.last->hand & USED_FLAG);
 
-				Balloon *new_last = h.last->prev;
```

Comment:
Declaration here was fine, no reason to reuse `b`

Reply:
Isn't reusing variables a good idea, to keep variable count low?
It may have an impact on readability, but in this case I don't think it has a significant effect, maybe having to look back to see `b` is a `Balloon *` type.

Reply:
A variable declaration within a loop makes it clear where it is used. A variable should be declared where it's used. Variable reuse is not a useful goal. This is not an optimisation

Reply:
I agree that it isn't an optimisation.
Having looked back at the function as a whole, I think using `b`, instead of introducing a new variable, makes more sense.

Reply:
The variable access within the loop doesn't use the initial value from outside or provide a value used after the loop. It is logically a different variable and was declared as such. There is no reason to change it. Variable reuse here is incorrect.

Reply:
I don't think they are that different.
If you wanted it to use the initial value within the loop this line could be rewritten as:
`b = b->prev`.
But `b = h.last->prev` is clearer in indicating the second last `Balloon` object.
Also, by reusing `b` in this way, it would be pointing to the last element, instead of a deleted memory location; after the for loop is done.

Reply:
Variables should be declared as late as possible and live only as long as they are needed. This is simple defensive c++. The existing version is more robust

Reply:
If you cared that much about variable scope, you'd end up putting code blocks around every variable. But that would make reading it near impossible.
How can the existing version be more robust?
If I tried to access `b` after the loop, I'd be trying to access an invalid memory location, the change prevents that.